### PR TITLE
Migrated project from pages to app router

### PR DESCRIPTION
**Pull request by zoosphar**
**Objective** 
This PR will migrate the jsoncrack project from the old next.js pages router to the new app router.

**Changes**
1. Project structure to consider the `app` directory as the entry point and move all the existing routes from the `pages` directory.
2. `pages/_app.tsx` and `pages/_document.tsx` files are replaced by `app/layout.tsx`.
3. Custom metatags inside the headtags are replaced by exporting the `metadata` constant from the `app/layout.tsx` file. I will continue doing the same for all files.
4. `pages/_error.tsx` is replaced by `app/not-found.tsx` which can not only serve a custom error UI but can also send HTTP status and serve metatags out of the box.
5. `next/router` is replaced by `next/navigation` which is a simplified API.

**References used for migration**
1. https://nextjs.org/docs
2. https://nextjs.org/docs/pages/building-your-application/upgrading/app-router-migration#migrating-from-pages-to-app

**Testing**
Tested the routes and functionalities by doing feature by feature test taking the production(jsoncrack.com) as the reference.

### CODE REVIEW

File: src/app/layout.tsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import StyledComponentsRegistry from "src/lib/registry";
+import CustomAppInit from "./CustomAppInit";
+
+export const metadata: Metadata = {
+  manifest: "/manifest.json",
+  icons: {
+    icon: "/favicon.ico",
+  },
+  description:
+    "Innovative and open-source visualization application that transforms various data formats, such as JSON, YAML, XML, CSV and more, into interactive graphs.",
```
Comment (TallTed) [inline src/app/layout.tsx:12]:
```suggestion
    "Innovative and open-source visualization application that produces interactive graphs from various data formats, such as JSON, YAML, XML, CSV and more.",
```

File: src/app/layout.tsx [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  openGraph: {
+    title: "JSON Crack - Visualize Data to Graphs",
```
Comment (TallTed) [inline src/app/layout.tsx:14]:
```suggestion
    title: "JSON Crack - Visualize Data as Graphs",
```

File: src/app/layout.tsx [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      },
+    ],
+    type: "website",
+    url: "https://jsoncrack.com",
+  },
+  twitter: {
+    card: "summary_large_image",
+    title: "JSON Crack - Visualize Data to Graphs",
+    description:
+      "Innovative and open-source visualization application that transforms various data formats, such as JSON, YAML, XML, CSV and more, into interactive graphs.",
```
Comment (TallTed) [inline src/app/layout.tsx:27]:
```suggestion
    title: "JSON Crack - Visualize Data as Graphs",
    description:
      "Innovative and open-source visualization application that produces interactive graphs from various data formats, such as JSON, YAML, XML, CSV and more.",
```

**Review by TallTed — CHANGES_REQUESTED** [review]
Language improvements

File: src/app/layout.tsx (hunk 1 above)
Comment (zoosphar) [reply src/app/layout.tsx:13]: @TallTed Thank you for suggesting these changes but I didn't change any language in the metatags, it was as it is from the original repo.
This PR only focuses on upgrading the project from the pages router to the app router.

Will fix these in another PR maybe.

**Comment by zoosphar** [comment]
> I'm not sure what you would like me to review. My previously suggested changes appear not to have been applied, though at least some of the text I suggested to change has been edited differently, but these edits are only visible commit-by-commit, not in the "[files changed](https://github.com/AykutSarac/jsoncrack.com/pull/380/files)"...

Hey @TallTed the changes you suggested are not in the scope of this PR, this PR only focuses on migrating the project from the pages router to the app router. Hence, I migrated the metadata from independent tags to a default object export.
Can you please review the migration and the adoption of new hooks from 'next/navigation' for routing.

**Comment by zoosphar** [comment]
Hey @TallTed 
I have updated the metadata in this PR only. Please have a look at this [commit](https://github.com/AykutSarac/jsoncrack.com/pull/380/commits/a055738dbf909c6500aac384f92446ba5c406bdf)

File: src/app/layout.tsx [hunk 4]
```diff
@@ -25,7 +25,7 @@ export const metadata: Metadata = {
     card: "summary_large_image",
     title: "JSON Crack - Visualize Data to Graphs",
```
Comment (TallTed) [inline src/app/layout.tsx:26]:
```suggestion
    title: "JSON Crack - Visualize Data as Graphs",
```

**Review by TallTed — COMMENTED** [review]
lgtm fwiw (I code almost exclusively in English)

**Comment by AykutSarac** [comment]
Good job! I've migrated the latest changes from the main branch. There are some console errors though, could you address them?

**Comment by AykutSarac** [comment]
Additionally theming is not working well, maybe we can improve it with this PR. The goal is to set Mantine & StyledComponents theme to light on pages except Editor & Widget, and for these 2 pages it'll be changed dynamically within the editor.

**Comment by AykutSarac** [comment]
Looks like no migration made for useSearchParams, can we do that? Also why did we add the "impl" package to the project?

File: src/app/docs/page.tsx [hunk 5]
```diff
@@ -36,14 +36,14 @@ const StyledHighlight = styled.span<{ $link?: boolean; $alert?: boolean }>`
   margin: ${({ $alert }) => ($alert ? "8px 0" : "1px")};
 `;
 
-const Docs = () => {
+export default function Page() {
   return (
     <Layout>
       <Head>
         <title>Documentation - JSON Crack</title>
         <meta name="description" content="Embedding JSON Crack tutorial into your websites." />
```
Comment (TallTed) [inline src/app/docs/page.tsx:44]:
```suggestion
        <meta name="description" content="Embedding JSON Crack tutorial in your websites." />
```

File: src/app/not-found.tsx [hunk 6]
```diff
@@ -1,45 +1,32 @@
...
+        <meta name="robots" content="noindex,nofollow" />
+      </Head>
+      <Stack mt={100} justify="center" align="center">
+        <Title fz={150} style={{ fontFamily: "monospace" }}>
+          404
+        </Title>
+        <Title order={2}>Nothing to see here</Title>
+        <Text c="dimmed" maw={800} style={{ textAlign: "center" }}>
+          Page you are trying to open does not exist. You may have mistyped the address, or the page
+          has been moved to another URL. If you think this is an error contact support.
```
Comment (TallTed) [inline src/app/not-found.tsx:22]:
```suggestion
          The page you are trying to open does not exist. You may have mistyped the address, or the page
          may have been moved to another URL. If you think this is an error, contact support.
```

**Comment by zoosphar** [comment]
> Looks like no migration made for useSearchParams, can we do that? Also why did we add the "impl" package to the project?

Hey @AykutSarac Thanks for the review, I have used useSearchParams from the next/navigation package which is supported in the app router, let me know why you think I should migrate it. 
And I have installed "impl" package because of the following error:
Screenshot 2024-02-11 at 9 57 28 PM

**Comment by zoosphar** [comment]
> It's recommended by Next.js, also [fails the build](https://github.com/AykutSarac/jsoncrack.com/actions/runs/7766105223/job/21181558599?pr=380). https://nextjs.org/docs/messages/missing-suspense-with-csr-bailout

Okay, got it. So adding a suspense boundary to the component would suffice right?

**Pull request closed without merging**